### DISCUSSION TITLE
면접 대비 질문 정리

### DISCUSSION THREAD

**Discussion by yoondgu** [woowacourse-study/2023-tal-study#58] [Announcements]
### 네트워크 계층
- OSI 7계층
- Frame, Packet, Segment, Datagram의 차이
  
### 프로토콜
- TCP/IP의 개념
  - TCP의 특징
  - TCP와 UDP의 차이
  - TCP와 UDP의 헤더 분석
- TCP의 3-way-handshake와 4-way-handshake
  - TCP의 연결 설정 과정(3단계)과 연결 종료 과정(4단계)이 단계가 - 차이나는 이유?
  - 만약 Server에서 FIN 플래그를 전송하기 전에 전송한 패킷이 Routing - 지연이나 패킷 유실로 인한 재전송 등으로 인해 FIN 패킷보다 늦게 도착하는 - 상황이 발생하면 어떻게 될까?
  - 초기 Sequence Number인 ISN을 0부터 시작하지 않고 난수를 생성해서 - 설정하는 이유?
- TCP 연결 끊김 탐지
### 웹
- www.naver.com 까지의 접속 과정
- HTTP의 특징
- HTTP와 HTTPS의 차이
- SSL
  - 공개키 알고리즘 방식
- HTTP와 HTTPS 동작 과정
- CORS란
- GET 메서드와 POST 메서드의 차이
- 쿠키(Cookie)와 세션(Session)
- DNS
- REST와 RESTful의 개념
### 기타
- 로드 밸런싱
- 서브네팅이란?
- 프록시 서버의 역할
- 소켓(Socket)이란
- Socket.io와 WebSocket의 차이
- unicast, multicast, broadcast 설명

> 출처 : 아래 두 링크의 질문을 재구성
> https://github.com/WeareSoft/tech-interview#2-network
> https://kjsu0209.github.io/Tech-Interview/network/network

**Comment by yeonkkk** [comment]
그저 빛 도이..✨

**Comment by yeonkkk** [comment]
중복되는 질문들도 있겠지만, 일단 저도 찾아서 올려봅니다.

## 네트워크 흐름
- [ ] 네트워크란?

##  OSI 7계층
- [ ] OSI 7계층이란 무엇이고 왜 계층으로 나눈 것인가요?
- [ ]  OSI 7계층의 각 특징을 설명해주세요.

### 전송 계층
- [ ]  전송계층의 역할은 무엇인가요?
- [ ]  TCP란?
- [ ]  3way handshake이란?
- [ ]  4way handshake이란?
- [ ]  3way handshake와 4way handshake가 차이나는 이유는?
- [ ]  TCP의 흐름제어, 혼잡제어란?
- [ ]  흐름제어기법 중 슬라이딩 윈도우 방식은 무엇인가요?
- [ ]  TCP의 오류제어 방법인 ARQ에 대해 설명해주세요.
- [ ]  TCP 헤더에는 어떤 정보들이 있나요?
- [ ]  UDP란?
- [ ]  TCP/IP 프로토콜 스택 4계층으로 구분짓고 설명해보세요.
- [ ]  tcp udp 패킷구조 차이점에 대해 설명해주세요.

### 네트워크 계층
- [ ]  패킷교환 방식에 대해 설명해주세요
- [ ] IP주소는 어떻게 구성되어 있나요?
- [ ] IP(IPv4)주소는 고갈되었다. 이를 해결하기 위한 방법이 있을까요?
- [ ]  IPv4와 IPv6에 대해 설명해주세요
- [ ]  공인IP와 사설IP에 대해서 설명해주세요
- [ ]  NAT기술에 대해 설명해주세요
- [ ]  고정 IP와 유동 IP에 대해서 설명해 주세요


### 응용 계층
- [x]  http와 https에 대해서 설명해주세요
- [ ]  http 1.1과 비교해 2.0의 특징에 대해서 설명해주세요
- [ ]  http 응답코드 종류에 설명해주세요
- [ ]  http 메서드 종류에는 뭐가 있을까요
- [x]  GET과 POST의 차이는?
- [x] PATCH과 PUT의 차이는?
- [ ]  DHCP란?
- [ ]  DNS란?

## 웹
- [ ]  REST API가 무엇인가요?
- [x]  HTTPS를 적용해서 통신하는 과정에 대해 설명해주세요
- [x] http keep alive / tcp keep alive
- [ ]  웹 서버 소프트웨어(Apache, Nginx)는 OSI 7계층 중 어디서 작동하는지 설명해보세요.
- [ ]  웹 서버 소프트웨어(Apache, Nginx)의 서버 간 라우팅 기능은 OSI 7계층 중 어디서 작동하는지 설명해보세요.
- [ ]  사용자가 웹 브라우저를 통해 서버에 이미지를 요청해서 사용자에게 보여주기까지 과정을 설명하세요.
- [ ]  로드 밸런싱이 무엇인지 설명해주세요
- [ ]  로드 밸런서가 서버를 선택하는 방식에 대해 설명해주세요.
- [ ]  blocking과 non-blocking에 대해 설명해주세요.
- [ ]  Synchronous와 Asynchronous에 대해 설명해주세요. 
- [x] CORS 란?

## 인증 & 보안
- [ ]  쿠키와 세션의 필요성에 대해 설명해주세요.
- [x]  Session과 Cookie 차이는?
- [ ]  Session 인증 방식과 Token 인증 방식의 차이점을 설명해주세요.
- [ ]  대칭키와 공개키에 대해 설명해주세요.

## 참고 자료
- [Egolper/beginners-tech-interview](https://github.com/Egolper/beginners-tech-interview/tree/main/network)
- [ham-study/cs-study-for-interview](https://github.com/ham-study/cs-study-for-interview/tree/main/Network)
- [devham76/tech-interview-study](https://github.com/devham76/tech-interview-study#2-network)
- [ksungdong/backend-interview-question](https://github.com/ksundong/backend-interview-question)
- [brave-people/brave-tech-interview](https://github.com/brave-people/brave-tech-interview/blob/main/contents/network.md)
- [gyoogle/tech-interview-for-developer](https://github.com/gyoogle/tech-interview-for-developer)
